tests/common/type_list.h

```diff
@@ -33,37 +33,137 @@ struct user_struct {
   }
 };
 
+namespace user_def_types {
+// A user-defined struct with several scalar member variables, no constructor,
+//  destructor or member functions.
+struct no_cnstr {
+  float a;
+  int b;
+  char c;
+
+  void operator=(const int &v) {
+    this->a = v;
+    this->b = v;
+    this->c = v;
+  }
+
+  friend bool operator==(const no_cnstr &lhs, const no_cnstr &rhs) {
+    return ((lhs.a == rhs.a) && (lhs.b == rhs.b) && (lhs.c == rhs.c));
+  }
+};
+
+// A user-defined class with several scalar member variables, a user-defined
+//  default constructor, and some member functions that modify the member
+//  variables.
+struct def_cnstr {
+  float a;
+  int b;
+  char c;
+
+ public:
+  constexpr def_cnstr() : a(3.0), b(2), c('c') {}
+
+  constexpr void assign(int val) {
+    a = val * 3.0;
+    b = val * 2;
+    c = val;
+  }
+
+  void operator=(const int &v) {
+    this->a = v * 3.0;
+    this->b = v * 2;
+    this->c = v;
+  }
+
+  inline friend bool operator==(const def_cnstr &lhs, const def_cnstr &rhs) {
+    return ((lhs.a == rhs.a) && (lhs.b == rhs.b) && (lhs.c == rhs.c));
+  }
+};
+
+// A user-defined class with several scalar member variables, a deleted default
+// constructor, and a user-defined (non-default) constructor.
+class no_def_cnstr {
+  float a;
+  int b;
+  char c;
+
+ public:
+  no_def_cnstr() = delete;
+
+  constexpr no_def_cnstr(int val) : a(val * 3.0), b(val * 2), c(val) {}
+
+  friend bool operator==(const no_def_cnstr &lhs, const no_def_cnstr &rhs) {
+    return ((lhs.a == rhs.a) && (lhs.b == rhs.b) && (lhs.c == rhs.c));
+  }
+
+  void operator=(const int &v) {
+    no_def_cnstr temp(v);
+    this->a = temp.a;
+    this->b = temp.b;
+    this->c = temp.c;
+  }
```

Comment:
```suggestion
  auto& operator=(const int &v) {
    return *this = no_def_cnstr { v };
  }
```
but it might be too terse?

Reply:
Actually you do not even need this at all since you have `constexpr no_def_cnstr(int val) : a(val * 3.0), b(val * 2), c(val) {}` which is implicit, so the copy constructor will be called afterward.
There might need some copy constructor to be defaulted to have this working if this is not already the case.